exonum/src/runtime/dispatcher/schema.rs

```diff
@@ -14,194 +14,239 @@
 
 //! Information schema for the runtime dispatcher.
 
+use exonum_crypto::Hash;
 use exonum_merkledb::{
-    access::{Access, AccessExt},
-    Fork, MapIndex,
+    access::{Access, FromAccess},
+    Fork, ListIndex, MapIndex, ObjectHash, ProofMapIndex,
 };
 
-use super::{ArtifactId, ArtifactSpec, Error, InstanceSpec};
-use crate::runtime::{DeployStatus, InstanceId, InstanceQuery};
+use crate::runtime::{InstanceId, InstanceQuery};
 
-const ARTIFACTS: &str = "core.dispatcher.artifacts";
-const PENDING_ARTIFACTS: &str = "core.dispatcher.pending_artifacts";
-const SERVICE_INSTANCES: &str = "core.dispatcher.service_instances";
-const PENDING_INSTANCES: &str = "core.dispatcher.pending_service_instances";
-const INSTANCE_IDS: &str = "core.dispatcher.service_instance_ids";
-const PENDING_INSTANCE_IDS: &str = "core.dispatcher.pending_instance_ids";
+use super::{
+    types::{ArtifactState, ArtifactStatus, InstanceState, InstanceStatus},
+    ArtifactSpec, Error, InstanceSpec,
+};
+
+const ARTIFACTS: &str = "dispatcher_artifacts";
+const PENDING_ARTIFACTS: &str = "dispatcher_pending_artifacts";
+const INSTANCES: &str = "dispatcher_instances";
+const PENDING_INSTANCES: &str = "dispatcher_pending_instances";
+const INSTANCE_IDS: &str = "dispatcher_instance_ids";
 
 /// Schema of the dispatcher, used to store information about pending artifacts / service
 /// instances, and to reload artifacts / instances on node restart.
 // TODO: Add information about implemented interfaces [ECR-3747]
-#[derive(Debug, Clone)]
-pub struct Schema<T> {
-    access: T,
+#[derive(Debug)]
+pub struct Schema<T: Access> {
```

Comment:
It could make sense to check there's no performance regression due to the eagerly initialized schema.